File: solutions/r/queen-attack/1/queen-attack.R
create <- function(row, col) {
  stopifnot(row>=0 & row<=7 & col>=0 & col<=7)
  c(row,col)
}

can_attack <- function(queen1, queen2) {
  (queen1[1]==queen2[1] |
  queen1[2]==queen2[2] |
  abs(queen1[1]-queen2[1])==abs(queen1[2]-queen2[2]))  
}

File: solutions/dart/queen-attack/1/lib/queen_attack.dart
class Queen {

  int row;
  int col;
  Queen(this.row, this.col) : assert(col>=0, 'column not positive'), assert(row>=0, 'row not positive'), assert(col<=7, 'column not on board'), assert(row<=7, 'row not on board');

  bool canAttack(Queen other){
    var dx = (this.col - other.col).abs();
    var dy = (this.row - other.row).abs();
    return (dx ==0) || (dy == 0) || (dx == dy);
  }  
}
